# This script calcutes energy from waveform file

## version 0.2a

In [6]:
import pandas as pd
import numpy as np
import configparser as cp

### Configure the lines below with filenames

In [7]:
file_cfg_in = '../dados/aquisicoes/Testes_10Fev2021/TesteNa22/central/config.ini'         # config.ini file input

file_in = ('../dados/aquisicoes/Testes_10Fev2021/TesteNa22/central/Waveform_0_0_1.txt',\
'../dados/aquisicoes/Testes_10Fev2021/TesteNa22/central/Waveform_0_1_1.txt',\
'../dados/aquisicoes/Testes_10Fev2021/TesteNa22/central/Waveform_0_2_1.txt',\
'../dados/aquisicoes/Testes_10Fev2021/TesteNa22/central/Waveform_0_3_1.txt',\
'../dados/aquisicoes/Testes_10Fev2021/TesteNa22/central/Waveform_0_4_1.txt',\
'../dados/aquisicoes/Testes_10Fev2021/TesteNa22/central/Waveform_0_5_1.txt',\
'../dados/aquisicoes/Testes_10Fev2021/TesteNa22/central/Waveform_0_6_1.txt',\
'../dados/aquisicoes/Testes_10Fev2021/TesteNa22/central/Waveform_0_7_1.txt',\
'../dados/aquisicoes/Testes_10Fev2021/TesteNa22/central/Waveform_1_0_1.txt',\
'../dados/aquisicoes/Testes_10Fev2021/TesteNa22/central/Waveform_1_1_1.txt',\
'../dados/aquisicoes/Testes_10Fev2021/TesteNa22/central/Waveform_1_2_1.txt',\
'../dados/aquisicoes/Testes_10Fev2021/TesteNa22/central/Waveform_1_3_1.txt',\
'../dados/aquisicoes/Testes_10Fev2021/TesteNa22/central/Waveform_1_4_1.txt',\
'../dados/aquisicoes/Testes_10Fev2021/TesteNa22/central/Waveform_1_5_1.txt',\
'../dados/aquisicoes/Testes_10Fev2021/TesteNa22/central/Waveform_1_6_1.txt',\
'../dados/aquisicoes/Testes_10Fev2021/TesteNa22/central/Waveform_1_7_1.txt')
file_out = ('../dados/resultados/Testes_10Fev2021/TesteNa22/central/py_charge_0_0.txt',\
'../dados/resultados/Testes_10Fev2021/TesteNa22/central/py_charge_0_1.txt',\
'../dados/resultados/Testes_10Fev2021/TesteNa22/central/py_charge_0_2.txt',\
'../dados/resultados/Testes_10Fev2021/TesteNa22/central/py_charge_0_3.txt',\
'../dados/resultados/Testes_10Fev2021/TesteNa22/central/py_charge_0_4.txt',\
'../dados/resultados/Testes_10Fev2021/TesteNa22/central/py_charge_0_5.txt',\
'../dados/resultados/Testes_10Fev2021/TesteNa22/central/py_charge_0_6.txt',\
'../dados/resultados/Testes_10Fev2021/TesteNa22/central/py_charge_0_7.txt',\
'../dados/resultados/Testes_10Fev2021/TesteNa22/central/py_charge_1_0.txt',\
'../dados/resultados/Testes_10Fev2021/TesteNa22/central/py_charge_1_1.txt',\
'../dados/resultados/Testes_10Fev2021/TesteNa22/central/py_charge_1_2.txt',\
'../dados/resultados/Testes_10Fev2021/TesteNa22/central/py_charge_1_3.txt',\
'../dados/resultados/Testes_10Fev2021/TesteNa22/central/py_charge_1_4.txt',\
'../dados/resultados/Testes_10Fev2021/TesteNa22/central/py_charge_1_5.txt',\
'../dados/resultados/Testes_10Fev2021/TesteNa22/central/py_charge_1_6.txt',\
'../dados/resultados/Testes_10Fev2021/TesteNa22/central/py_charge_1_7.txt')
file_out_1 = ('../dados/resultados/Testes_10Fev2021/TesteNa22/central/Waveform_0_0_2.txt',\
'../dados/resultados/Testes_10Fev2021/TesteNa22/central/Waveform_0_1_2.txt',\
'../dados/resultados/Testes_10Fev2021/TesteNa22/central/Waveform_0_2_2.txt',\
'../dados/resultados/Testes_10Fev2021/TesteNa22/central/Waveform_0_3_2.txt',\
'../dados/resultados/Testes_10Fev2021/TesteNa22/central/Waveform_0_4_2.txt',\
'../dados/resultados/Testes_10Fev2021/TesteNa22/central/Waveform_0_5_2.txt',\
'../dados/resultados/Testes_10Fev2021/TesteNa22/central/Waveform_0_6_2.txt',\
'../dados/resultados/Testes_10Fev2021/TesteNa22/central/Waveform_0_7_2.txt',\
'../dados/resultados/Testes_10Fev2021/TesteNa22/central/Waveform_1_0_2.txt',\
'../dados/resultados/Testes_10Fev2021/TesteNa22/central/Waveform_1_1_2.txt',\
'../dados/resultados/Testes_10Fev2021/TesteNa22/central/Waveform_1_2_2.txt',\
'../dados/resultados/Testes_10Fev2021/TesteNa22/central/Waveform_1_3_2.txt',\
'../dados/resultados/Testes_10Fev2021/TesteNa22/central/Waveform_1_4_2.txt',\
'../dados/resultados/Testes_10Fev2021/TesteNa22/central/Waveform_1_5_2.txt',\
'../dados/resultados/Testes_10Fev2021/TesteNa22/central/Waveform_1_6_2.txt',\
'../dados/resultados/Testes_10Fev2021/TesteNa22/central/Waveform_1_7_2.txt')

### Read config.ini

In [8]:
## Read config.ini file and get the variables
cfg = cp.RawConfigParser()
cfg.read(file_cfg_in)

#cfg.get('DPPCI','MAXNB')

par = dict(cfg.items('DPPCI'))

for p in par:
    par[p]=par[p].split("#",1)[0].strip() # To get rid of inline comments

globals().update(par)  #Make them availible globally

print('threshold: ', thr)
thr = int(thr)
print('record length: ', recordlength)
recordlength = int(recordlength)

options = {
    '0': 0,
    '1': 16,
    '2': 32,
    '3': 128}
nsbl = options.get(nsbl, 'invalid data')
print('nsbl: ', nsbl)
nsbl = int(nsbl)
print('pre-gate: ', pgate)
pgate = int(pgate)
print('gate: ', gate)
gate = int(gate)
print('pulse polarity: ', pulsepolarity, ' | 0 = positive; 1 = negative')
print('analog trace 2: ', analog_trace_2, ' | 5 = baseline; 6 = threshold; 8 = none (dual trace off) | from waveform_b_c_2.txt')
pretriggersize = int(pretriggersize)
print('pre-trigger size: ', pretriggersize)

nevt = len(pd.read_csv(file_in[0], header = None, sep = '\t')) #Take events grom Waveform_0_0_1.txt
print('number of events: ', nevt)

threshold:  45
record length:  150
nsbl:  32
pre-gate:  55
gate:  90
pulse polarity:  1  | 0 = positive; 1 = negative
analog trace 2:  8  | 5 = baseline; 6 = threshold; 8 = none (dual trace off) | from waveform_b_c_2.txt
pre-trigger size:  65
number of events:  6296


### If the config values in config.ini were wrong, force with the values below
#### Example:
<ul>
<li>threshold:  45</li>
<li>record length:  150</li>
<li>nsbl:  32</li>
<li>pre-gate:  55</li>
<li>gate:  90</li>
<li>pulse polarity:  1  | 0 = positive; 1 = negative</li>
<li>analog trace 2:  8  | 5 = baseline; 6 = threshold; 8 = none (dual trace off) | from waveform_b_c_2.txt</li>
<li>pre-trigger size:  65</li>
<li>number of events:  6296</li>
</ul>

<img src="./energia-16canais-pulso.png">
<img src="./energia-16canais-pulso16.png">


There is a delay for board 0 (channels 0 to 7) and another for board 1 (channels 8 to 15).

In [9]:
## Add these lines only if necessary to force the values:

#nevt = 7442
# analog_trace_2 = '8'
recordlength = 144
nsbl = 5
shift = 0 # number of samples to shift left to calculate the baseline
nchannels = 16 # number of channels

### Process the 16 pulses and calculates energy

<img src="files/energia-16canais-pulsos.png">

The baseline is unconsistent due the different delay values for each board.<br>
In this specific case, is necessary to re-calculate the baseline for a better value.

In [5]:
## Processing pulses
for k in tqdm(range(0, nchannels)):
    val_chg = np.zeros(nevt, dtype = int) 
    val_bsl = np.zeros((nevt, recordlength), int)    
    val_pls = pd.read_csv(file_in[k], header = None, sep = '\t', nrows=nevt)
    val_pls = val_pls.to_numpy()
    for j in tqdm(range(0, nevt)):  

        # Baseline
        # val_bsl (from file or from this algorithm)
        if (analog_trace_2 != '5'):    
            for i in range(0, recordlength):   
                if ((i > (recordlength - nsbl)) or (i < nsbl)):                                    
                    val_bsl[j, i] = val_pls[j, i]
                else:                
                    val_bsl[j, i] = round(np.sum(val_pls[j, (i - nsbl) : i]) / nsbl)       

        # Energy    
        for i in range(pretriggersize - pgate, pretriggersize + gate - pgate):         
            #print('Trg: {}, Pos: {}, Bsl-st: {}, Bsl-fi: {}'.format(pos_trg[j], i, val_bsl[j, pos_trg[j] - pgate], val_bsl[j, pos_trg[j] + gate - pgate]))
            if pulsepolarity == '0':
                bl = val_bsl[j, pretriggersize - pgate - shift]
#                 if val_pls[j, i] >= bl: # ignore negative samples
#                     val_chg[j] = val_chg[j] + (val_pls[j, i] - bl) #freeze baseline value in gate start  
                val_chg[j] = val_chg[j] + (val_pls[j, i] - bl) #freeze baseline value in gate start                                      
            else:
                bl = val_bsl[j, pretriggersize - pgate - shift]
#                 if val_pls[j, i] <= bl: # ignore positive samples
#                     val_chg[j] = val_chg[j] - (val_pls[j, i]  - bl)   
                val_chg[j] = val_chg[j] - (val_pls[j, i] - bl) #freeze baseline value in gate start      
    
    #Save charges into file  
    np.savetxt(file_out[k], abs(val_chg), fmt='%d') 

ImportError: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html